make `add` an alias of `update`

mitchellwrosen opened the pull request:
## Overview

This PR removes the old `add` behavior, and makes `add` an alias of `update`.

mitchellwrosen committed:
make add an alias for update, and update transcripts

mitchellwrosen committed:
remove collision

mitchellwrosen committed:
refactor

mitchellwrosen committed:
Revert "remove collision"

This reverts commit 559637471a02ac847f02924e91aa852492682ff3.

mitchellwrosen committed:
fix bug in refactor

mitchellwrosen commented on unison-src/transcripts/merge.output.md:
```diff
@@ -250,7 +250,7 @@ scratch/alice> view foo bar
 scratch/alice> display bar
 
-  "old foo - old foo"
+  "new foo - new foo"
```
This is a bit perplexing... the new output is correct. 😅

mitchellwrosen committed:
run transcripts-round-trip/main.md transcript manually

mitchellwrosen committed:
run manual transcripts

ChrisPenner commented on unison-src/transcripts/idempotent/fix-ls.md:
```diff
@@ -14,8 +14,7 @@ foo.bar.subtract x y = x Int.- y
-  change:
+  do an `update`, here's how your codebase would change:
 
     ⍟ These new definitions are ok to `add`:
```
Should probably update `add` -> `update` in this message too

mitchellwrosen replied:
Oops! Yeah

ChrisPenner commented on unison-src/transcripts/idempotent/fix-ls.md:
```diff
@@ -26,10 +25,10 @@ foo.bar.subtract x y = x Int.- y
-  ⍟ I've added these definitions:
+  Okay, I'm searching the branch for code that needs to be
+  updated...
 
-    foo.bar.add      : Int -> Int -> Int
-    foo.bar.subtract : Int -> Int -> Int
+  Done.
```
I notice `add` used to show what actually changed; is that something we should have in `update` too?

mitchellwrosen replied:
We certainly could!

aryairani replied:
We should! We can update the display template design and take a look after any already in progress `merge`/`update` stuff

aryairani replied:
What I mean is that we can take a look at what we should provide in a future slurp output vs a future `update` output.

ChrisPenner reviewed (approved):
Looks good, I left a note on a couple messages that maybe should change.

Also, maybe I just missed it, but we should probably change `add.run` to `update.run` or something like it :)

mitchellwrosen committed:
update transcripts

mitchellwrosen committed:
⅄ trunk → 25-05-15-delete-add

aryairani commented:
I'm gonna try to do a pass over this after releasing 0.5.41

mitchellwrosen committed:
⅄ trunk → 25-05-15-delete-add

aryairani commented on unison-src/transcripts-using-base/serialized-cases/case-04.v4.hash:
```diff
@@ -1 +1 @@
-EXAQLMU6IKGAY7DNOHND5VUQQAQPIJN3IVCF5DISOOEVLRQZ3Q2CZOYEVDMY7MYQX2CG6CJFH2HQD6XOMKHQNK5JUZB3G7RZQNREQRQ=
\ No newline at end of file
+KPXQRV7VO4LIVLHVTR5HXUDOXSFFF2C4QM7IMLCU3YZXJSTJSFDT7YUZCWEW4TLH4SHAHT7NHB2ZTUUWGRGUTWGBVY7BFA33XWRBDKQ=
```
Why do you think this changed?

mitchellwrosen replied:
Huh, I have no idea how those got regenerated... I reverted both

aryairani replied:
Huh. I wonder why it was passing both before and after.

mitchellwrosen replied:
It was two files that I regenerated somehow, the input and the output

aryairani replied:
This is still happening, I will work on bisecting.

mitchellwrosen replied:
What's still happening?

mitchellwrosen committed:
revert case-04 serialized cases

aryairani commented on unison-src/transcripts-manual/remote-tab-completion.output.md:
````diff
@@ -3,8 +3,6 @@
 ``` ucm
-.> debug.tab-complete pull unison.pub
-
-   unison.public
````
I guess this is an old change in a transcript that's run infrequently, but what happened here?

mitchellwrosen replied:
I'm actually not sure, it doesn't look related to these changes. I think I manually ran more transcripts in `transcripts-manual` than strictly necessary, just trying to hit all the random stuff that calls `add`... could revert or leave it, what do you think?

aryairani replied:
I'll take another look to see if this whole transcript is deprecated or what.  remote tab completion sounds like a "loose code" thing.

aryairani replied:
Ok this is literally the whole transcript.
@ChrisPenner Is the `remoteNamespaceArg` / `sharePathCompletion` alive in anyway? Running the tab completion just seems to give me a `BEL`.

aryairani commented on unison-src/transcripts-using-base/binary-encoding-nats.output.md:
````diff
@@ -77,18 +76,10 @@ testABunchOfNats _ =
 ``` ucm
 scratch/main> add
 
-  ⍟ I've added these definitions:
-
-    type EncDec
-    BE16             : EncDec
-    BE32             : EncDec
-    BE64             : EncDec
-    LE16             : EncDec
-    LE32             : EncDec
-    LE64             : EncDec
-    testABunchOfNats : ∀ _. _ ->{IO} [Result]
-    testNat          : Nat -> '{IO, Stream Result} ()
-    testRoundTrip    : Nat -> EncDec ->{IO, Stream Result} ()
````
I already somewhat miss this, but maybe the slurp result v2 will be enough, so we can try that first before worrying about replacing this functionality.

mitchellwrosen replied:
I agree it would be nicer to say something other than `Done` in the `update` feedback, but you do pretty much get this exact `add` output _before_ running `update`, so I guess that's part of the reason it hasn't been on our radar to make it more verbose.

But it'd be trivial to add back, plus more – you might like to see number of propagated updates, for example

aryairani replied:
Yeah agreed about the slurp output. I was imagining we could eventually put the number of propagated updates there too or instead.

aryairani commented on unison-src/transcripts/idempotent/delete.md:
```diff
@@ -239,11 +239,9 @@ scratch/main> delete.verbose a b c Foo
 scratch/main> delete.verbose Foo.Foo
 
-  Name changes:
+  Removed definitions:
 
-    Original      Changes
-    1. Foo.Foo ┐  2. Foo.Foo (removed)
-    3. foo.Foo ┘  
```
I wonder what's happening here?

aryairani replied:
Any ideas on this one? @mitchellwrosen

mitchellwrosen replied:
No clue, the old output doesn't look correct at all, but I don't know how the transcript was previously working.

mitchellwrosen replied:
Oh, I see what's going on, I think. The transcript is really long and sloppy, not cleaning up after itself between test cases. It's `add`ing an alias (`structural type Foo = Foo ()`) whereas in the new transcript, we `update` and thus update dependents of the type that already exists.

aryairani commented on unison-src/transcripts/idempotent/fuzzy-options.md:
````diff
@@ -61,11 +61,6 @@ scratch/main> debug.fuzzy-options view _
 Namespace args
 
 ``` ucm
-scratch/main> add
-
-  ⊡ Ignored previously added definitions: nested.optionTwo
-    optionOne
-
````
What's up with this one?

aryairani replied:
it's just two adds in a row for no reason

aryairani commented on unison-src/transcripts/idempotent/help.md:
```diff
@@ -930,20 +924,20 @@ scratch/main> help-topic filestatus
   needs update         A definition with the same name as an
-                       existing definition. Doing `update`
-                       instead of `add` will turn this failure
-                       into a successful update.
+                       existing definition. Doing `update` will
+                       turn this failure into a successful
+                       update.
```
What does "needs update" mean in the context of this PR?

aryairani replied:
I deleted "needs update"; it does show up in `add.run` though.

aryairani commented on unison-src/transcripts/idempotent/add-run.md:
we should take a look at `add inc` and see what this transcript is trying to demonstrate

aryairani replied:
On `trunk`, we get this:

```
scratch/main> add.run xres

  x These definitions failed:

    Reason
    needs update   xres   : Nat

    Tip: Use `help filestatus` to learn more.
 ```
 
 meaning that add.run fails because `xres` already exists in the branch.  We get a message saying "needs update" which is not meaningful in the context of `add.run`, and also not a valid outcome in any case if we are replacing `add` with `update`.

mitchellwrosen replied:
Yeah, `add.run` is a little weird so I just left it alone in this PR. But for that reason, I think it is a valid outcome (well, as valid as before) to say "needs update".

aryairani replied:
Hm, but there's no way to "update" currently in the context of `.run`. (I'm not convinced we should automatically be overwriting it either; there's no "slurp" phase for `add.run` to confirm your intentions.)

Depending on how difficult it is, I'm thinking we should have a different output message that says to rename or delete `xres` and try again — or, if we gain a "run-and-update" command, then we could recommend that instead.

aryairani replied:
Also because I was trying to delete "needs update" since it doesn't really exist; I guess I could revert that for now though.

aryairani replied:
I reverted deleting needs update, and updated the message to fit `add.run`.

aryairani commented on unison-src/transcripts/idempotent/block-on-required-update.md:
````diff
@@ -59,11 +58,11 @@ Try to add only the new `y`. This should fail because it requires an update to `
 ``` ucm :error
 scratch/main> add y
````
we should take a look at this `add y` and determine what this transcript is trying to show

aryairani replied:
This file goes away, because we no longer distinguish adds vs updates.

aryairani commented on unison-src/transcripts/idempotent/input-parse-errors.md:
```diff
@@ -38,6 +33,13 @@ scratch/main> ls .
 
 scratch/main> add 1
```
what are we trying to show here with the numbered args

aryairani replied:
I removed the section that demos `add` since we deleted it as a distinct command.

aryairani commented on unison-src/transcripts/project-outputs/.github/ISSUE_TEMPLATE/bug_report.output.md:
````diff
@@ -15,11 +15,19 @@ This should be written as a [ucm transcript](https://www.unison-lang.org/docs/to
 ```
 
-Here I typo the next command and `ucm` silently does nothing, I would have expected an error message:
+Here I try to pass an argument to `update`, which fails:
 
-``` ucm
-scratch/main> add b
````
- [ ] fix bug report template

aryairani replied:
I'm leaving it alone because it's still a reasonable template even though it exemplifies one that would be closed as wontfix.

aryairani commented on unison-src/transcripts/project-outputs/docs/type-declarations.output.md:
probably shouldn't mess with this one

aryairani commented on unison-src/transcripts/project-outputs/.github/ISSUE_TEMPLATE/bug_report.output.md:
````diff
@@ -15,11 +15,19 @@ This should be written as a [ucm transcript](https://www.unison-lang.org/docs/to
-Here I typo the next command and `ucm` silently does nothing, I would have expected an error message:
+Here I try to pass an argument to `update`, which fails:
 
-``` ucm
-scratch/main> add b
````
fix up the bug report template

aryairani replied:
disregard; it's alright.

aryairani commented on unison-src/transcripts/project-outputs/docs/publishing-library1.output.md:
probably just revert this file too

aryairani replied:
reverted

aryairani commented on unison-src/transcripts/idempotent/input-parse-errors.md:
need to take a deeper look at this file

aryairani commented on unison-src/transcripts-using-base/test-watch-dependencies.md:
check this one

aryairani replied:
ended up deleting it, as the only interesting test AFAICT was the one testing the selective `add`, which no longer exists.

aryairani committed:
transcript cleanup from PR review

- reverted docs/; these historical notes don't need `add` replaced with `update`
- update the help that gets displayed on an `add.run` conflict
- consider `debug.tab-complete` an `:error` when there are no available completions
- delete `block-on-required-update.md` since that only applied to the old `add` command
- delete `test-watch-dependencies.md` since that only applied to selective `add`
- delete the parse errors demo for selective `add` in `input-parse-errors.md`
- update and move `merge.md` into idempotent/
- update `remote-tab-completion.md` to mention that it's apparently not working.
  - we could move this to idempotent I think, even though it hits the network.
- update `case-04.v4.ser` per https://github.com/unisonweb/unison/issues/5726
